[Doc Comment Translation] Improve Remarks Formatting

jsquire opened the pull request:
# Summary

The focus of these changes is to improve formatting for the `<remarks>` element of the XML doc comments to treat them as text rather than pre-formatted blocks.

Previously, transformation was sensitive to the formatting of the source code, attempting to preserve line breaks and spacing rather than treating it as XML content.  The source code was often formatted to wrap lines so that they could be more easily read by other developers, causing the rendering for the MS Docs site to contain awkward and undesirable formatting.

With these changes, the `<remarks>` are treated as content, normalizing line breaks and spacing and introducing explicit paragraphs to ensure that the structure is semantically accurate and assist in HTML rendering.

This should address the asks in DevOps [#136807](https://ceapex.visualstudio.com/Engineering/_workitems/edit/136807) and  [#373096](https://ceapex.visualstudio.com/Engineering/_workitems/edit/373096).  It should also address the other remarks formatting issues that have been reported to the Azure SDK repository, such as the `EventDataBatch::TryAdd` member:

![image](https://user-images.githubusercontent.com/913445/116317480-60b49800-a781-11eb-9299-fef2e83b0c8a.png)

![image](https://user-images.githubusercontent.com/913445/116317432-5397a900-a781-11eb-96e9-e1e0ecba76fa.png)

# References and Related

- [Azure DevOps #136807](https://ceapex.visualstudio.com/Engineering/_workitems/edit/136807) _(Microsoft Internal)_
- [Azure DevOps #373096](https://ceapex.visualstudio.com/Engineering/_workitems/edit/373096) _(Microsoft Internal)_
- [[Service Bus] Docs remarks layout is Incorrect (#18783)](https://github.com/Azure/azure-sdk-for-net/issues/18783)
- [Markup mistake in Remarks section #18025](https://github.com/Azure/azure-sdk-for-net/issues/18025)
- [[DOCS] Remarks have strange formatting in API docs for keyclient.getkeyasync (#13567)](https://github.com/Azure/azure-sdk-for-net/issues/13567)

jsquire committed:
[Doc Comment Translation] Improve Remarks Formatting

The focus of these changes is to improve formatting for the `<remarks>`
element of the XML doc comments to treat them as text rather than
pre-formatted blocks.

Previously, transformation was sensitive to the formatting of the source
code, attempting to preserve line breaks and spacing rather than treating
it as XML content.  The source code was often formatted to wrap lines so
that they could be more easily read by other developers, causing the
rendering for the MS Docs site to contain awkward and undesirable formatting.

With these changes, the `<remarks>` are treated as content, normalizing
line breaks and spacing and introducing explicit paragraphs to ensure that
the structure is semantically accurate and assist in HTML rendering.

This should address the ask in DevOps [#136807](https://ceapex.visualstudio.com/Engineering/_workitems/edit/136807)
as well as the choppy lines rendered for other Azure SDK types, such as the
[`EventDataBatch`](https://docs.microsoft.com/dotnet/api/azure.messaging.eventhubs.producer.eventdatabatch.tryadd?view=azure-dotnet#remarks)
class.

jsquire commented:
//cc: @DanielJurek, @AlexGhiondea, @TianqiZhang

jsquire commented on test/xml/CatLibrary/Cat`2.xml:
```diff
@@ -93,9 +93,9 @@
     <remarks>
```
@TianqiZhang:  I had some difficulty regenerating the XML using the `gen-xml.ps1` script in the `test` folder.  Running it introduced a lot of subtle formatting changes which upset the diff tool.   I suspect that it has to do with the .NET SDK versions.

TianqiZhang replied:
@jsquire thank you for your contribution! The diff is probably caused by the fact that sometimes people modified the xml directly, instead of changing the test assembly source code. I tested `gen-xml.ps1` locally and saw many modified lines in ``Cat`2.xml``. I'll fix the xml diff first so it won't be confusing.

TianqiZhang commented on ECMA2Yaml/ECMAHelper/ECMALoader.docs.cs:
```diff
@@ -124,7 +126,7 @@ public Docs LoadDocs(XElement dElement, string filePath)
             {
                 Summary = NormalizeDocsElement(dElement.Element("summary")),
-                Remarks = remarksText,
+                Remarks = FormatTextIntoParagraphs(remarksText),
```
Actually there are a lot of cases where writers use `<format type="text/markdown">` in `<remarks>` to tell ECMA2Yaml that they are using Markdown syntax, like this one: https://github.com/dotnet/dotnet-api-docs/blob/main/xml/System.Windows/SystemColors.xml#L4504. So how about call `FormatTextIntoParagraphs()` inside `NormalizeDocsElement`, so that we only do this step when it's not Markdown?

When it comes to a format change, I usually diff the output yamls before/after my change in two source data sets: https://github.com/dotnet/dotnet-api-docs/tree/main/xml and https://github.com/Azure/azure-docs-sdk-dotnet/tree/master/xml.

jsquire replied:
Thanks, @TianqiZhang.   That's helpful.   

There are definitely scenarios that I'm not accounting for due to lack of awareness.   To be transparent, I wasn't aware of  `<format>` -  it doesn't exist in the [C# Documentation Comments](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/documentation-comments) section of the language spec nor in the [triple slash comments](https://dotnet.github.io/docfx/spec/triple_slash_comments_spec.html) section of the DocFX spec.  If you've got any kind of spec or documentation about formatting, please feel free to point me at that so I can take it into consideration.   

If you don't mind, I'd like to talk through things a bit to capture any corner cases.  In the base case, we know that this pattern is one we want to ignore and leave as-is:

```xml
<remarks>
  <format type="text/markdown">
  <![CDATA[  

... MARKDOWN CONTENT ...
  
   ]]>
  </format>
</remarks>
```

What about a scenario where `<format>` is partial content within the `<remarks>` tag?  In the following example, if it is a valid case, I'd assume that we want to transform the first and third paragraphs and ignore the second due to `<format>` :

### BEFORE
```xml
<remarks>
  This is the first paragraph of text before we 
  drop down into some markdown formatting.

  <format type="text/markdown">
  <![CDATA[  

... MARKDOWN CONTENT ...
  
   ]]>
  </format>

  This is a trailing text block that represents a third
  paragraph that probably tells you wonderful things.
</remarks>
```

### AFTER
```xml
<remarks>
  <p>This is the first paragraph of text before we drop down into some markdown formatting.</p>
  <format type="text/markdown">
  <![CDATA[  

... MARKDOWN CONTENT ...
  
   ]]>
  </format>
  <p>This is a trailing text block that represents a third paragraph that probably tells you wonderful things.</p>
</remarks>
```



The next consideration is an embedded `<format>`.  Given the following:

### BEFORE
```xml
<remarks>
  We're just here, minding our own business when a
  wild <format type="text/markdown"><![CDATA[  **_APPEARS_**  ]]></format>.  What now?

  This is a trailing text block that represents a third
  paragraph that probably tells you wonderful things.
</remarks>
```

I think that we'd want the transform to look like:
_(wrapping included here for readability, it would all be a single line in the actual transform)_

### AFTER
```xml
<remarks>
  <p>We're just here, minding our own business when a wild <format type="text/markdown"><![CDATA[  **_APPEARS_**  ]]></format>. What now?</p>
  <p>This is a trailing text block that represents a third paragraph that probably tells you wonderful things.</p>
</remarks>
```

Any other tags beyond `<format>` that we should account for, or other scenarios that may influence the transform?   I'd appreciate any insight that you may have.

TianqiZhang replied:
I'm not sure if there's any official document about `<format>` usage in `<remarks>`, @joelmartinez do you know if there's a guideline that dotnet team is following when contributing to dotnet-api-docs repo?

Anyway, the only usage that works in ECMA2Yaml is to use `<format>` as the only child under `<remarks>`. So one way that I can think of is to call `FormatTextIntoParagraphs()` inside `NormalizeDocsElement` and only call it when  `<remarks>`'s first child is not `<format>`.

joelmartinez replied:
I'm not sure of any documentation ... this element was added at some point as part of the onboarding of mdoc to be used in docsmsft's API pipeline; the idea was that we wanted to support markdown, but wanted to make it an explicit opt-in instead of redefining formatting rules for plain text in the summary and other elements.

jsquire replied:
Thanks for the discussion and help, folks.   Sounds like the simplest short-term path forward will be to skip formatting when the `<remarks>` content starts with `<format` and ignore any embedding.   

Longer term, now that I know this tag exists, I want to build some additional intelligence into `FormatTextIntoParagraphs` to detect cases where it's not scoped to the entire `<remarks>` body and also recognize `<pre>`  as formatting that should be respected.   I've got some prototyping around this that I'm playing with, though I'll likely do that as a separate PR.

jsquire replied:
@TianqiZhang:  I ended up delegating to the existing logic in `NormalizeDocsElement` to detect the `<format>` wrapping and let it have complete ownership of normalization, skipping paragraph formatting.   I believe this was the direction that you were advocating for as well, but please correct me if I'm mistaken.

I added some test cases at the unit level as well as in the integration diffs to cover this;  there's some spacing that looks odd to me, but I didn't make changes since the logic was already in place.   I'll call them out inline, if you would be so kind as to offer thoughts.

TianqiZhang replied:
@jsquire thanks for the update, I'll diff the result by tomorrow and merge it if the result looks good in dotnet-api-docs repo.

TianqiZhang replied:
One more request @jsquire , sorry it took so long, I see there are a lot of diffs like below, which makes it difficult to review the real diffs. Is it possible to avoid touching the content if there's only 1 line?
![image](https://user-images.githubusercontent.com/5326582/118252800-bd32e980-b4db-11eb-8022-f4fc43179ea2.png)

jsquire replied:
Hi @TianqiZhang.   Thanks for your feedback.  I adjusted parsing to no longer wrap a single line when no embedded paragraph tags are detected.  Tests and diff assets have been updated to reflect the new behavior.

I also  took a look through a few different scenarios for the MS Docs site output, and it looks like a single line of text gets rendered in `<p>` tags without explicitly adding them - so the user experience shouldn't be negatively impacted by this change.

TianqiZhang replied:
Diffs looking good, change merged, thanks again @jsquire !

jsquire committed:
Ignore <remarks> wrapped in <format>

This set of changes updates processing for remarks to make use of the
existing logic in NormalizeDocsElement to detect the pattern:

<remarks>
  <format>
    ...
  </format>
</remarks>

When this pattern is detected, the remarks are considered preformatted
and parsing into paragraphs will be skipped.

jsquire commented on test/yml_SDP/CatLibrary.Cat-2.Age.yml:
````diff
@@ -27,6 +27,14 @@ members:
+    var age = cat.Age;  // Will always be 1.
+
+    ```
   metadata: {}
````
This spacing for lines was controlled by the existing logic for normalizing in `NormalizeDocsElement`.    To me, it looks wrong when considered against the source and the XML output:

```xml
       <remarks>
          <format langauge="markdown">
            ```csharp
            var cat = new Cat();
            var age = cat.Age;  // Will always be 1.
            ```
            </format>
        </remarks>
```

TianqiZhang replied:
The empty lines are actually added by yaml serializer automatically, it has its own minds and somehow chooses newline style for me, see details in https://yaml-multiline.info/. So don't worry about it, this is correct.

jsquire replied:
Thanks for the confirmation; just wanted to make sure that I didn't accidentally introduce weird behavior.

jsquire commented on test/yml_SDP/CatLibrary.Cat-2.ownEat.yml:
```diff
@@ -28,6 +28,13 @@ members:
+
+
+    Nor this one.
   metadata: {}
```
Here, again, this spacing was performed by the existing logic in `NormalizeDocsElement` but looks off to me based on the source and XML:

```xml
        <remarks>
        <format>
          This should not be changed.
          Not this line.

          Nor this one.
        </format>
        </remarks>
```

TianqiZhang replied:
Similar to the last one, this is just caused by yaml serializing style.

jsquire committed:
Don't Wrap Single Line In Paragraph Tags

Adjust the approach to not wrap text that is a single line without
embedded paragraphs with <p></p> tags; the MS Docs site will render
a single unwrapped line as a paragraph by default.

TianqiZhang commented:
This change could potentially fix [AB#136807](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/136807).
